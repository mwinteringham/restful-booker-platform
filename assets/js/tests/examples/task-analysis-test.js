import React from 'react';
import nock from 'nock';
import LoginComponent from '../../src/js/components/Login.js';
import { render, fireEvent, waitFor } from '@testing-library/react';

// We first declare our test by using Jests test function and provide it
// with a name and then an anonymous function which will run our test
test('Login component is created', () => {
    // Next we create our component that we want to check. By using
    // React Testing Libraries' render function and passing it the React
    // component we want to create, RTL will create a headless DOM and 
    // render the React component inside.
    const { asFragment } = render(<LoginComponent />);

    // We then compare what the component has done to a previously saved
    // state that lives in the __snapshots__ folder. If anything has changed
    // it will raise an alert.
    expect(asFragment()).toMatchSnapshot();
});

// We first declare our test by using Jests test function and provide it
// with a name and then an anonymous function which will run our test
test('Login component sends correct payload', async () => {
    // Our Login object is going to send an API request that we want to capture
    // and inspect to make sure the request body matches what we want. We do this
    // by programming a mock using nock with the expected URL and post body and 
    // finally create a fake response for it to reply with
    let authMock = nock('http://localhost')
        .post('/auth/login', {
            "username" : "admin",
            "password" : "password"
        })
        .reply(200, { token: '123ABC' });

    // Next we create our component that we want to check. By using
    // React Testing Libraries' render function and passing it the React
    // component we want to create, RTL will create a headless DOM and render
    // the React component inside.
    const {getByTestId} = render(
        <LoginComponent setAuthenticate={()=>{}} />,
    )

    // We next fill in the LoginComponent form fields and submit it
    fireEvent.change(getByTestId('username'), { target: { value: 'admin' } });
    fireEvent.change(getByTestId('password'), { target: { value: 'password' } });
    fireEvent.click(getByTestId('submit'));

    // Finally, we have to call isDone() on our mock to check whether
    // or not it received the expected request. If it did, it will return
    // true. If it didn't it will return false. We then assert whether 
    // the result is true.
    await waitFor(() => expect(authMock.isDone()).toBeTruthy())
});
