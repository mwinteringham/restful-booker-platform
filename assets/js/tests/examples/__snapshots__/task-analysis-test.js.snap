// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component is created 1`] = `
<DocumentFragment>
  <div
    style="margin-top: 15%;"
  >
    <div
      class="row"
    >
      <div
        class="col-sm-2"
      />
      <div
        class="col-sm-8"
        style="text-align: center;"
      >
        <h2
          data-testid="login-header"
        >
          Log into your account
        </h2>
        <br />
      </div>
      <div
        class="col-sm-2"
      />
    </div>
    <div
      class="row"
    >
      <div
        class="col-sm-4"
      />
      <div
        class="col-sm-4"
      >
        <div
          class="form-group"
        >
          <input
            class="form-control"
            data-testid="username"
            id="username"
            placeholder="Username"
            style="border: 1px solid grey;"
            type="text"
          />
        </div>
        <div
          class="form-group"
        >
          <input
            class="form-control"
            data-testid="password"
            id="password"
            placeholder="Password"
            style="border: 1px solid grey;"
            type="password"
          />
        </div>
        <button
          class="btn btn-primary float-right"
          data-testid="submit"
          id="doLogin"
          type="submit"
        >
          Login
        </button>
      </div>
      <div
        class="col-sm-4"
      />
    </div>
  </div>
</DocumentFragment>
`;
